### ISSUE TITLE
"map" support for IDL

### ISSUE THREAD

**Issue by JonathanWheelerL3** [OpenDDS/OpenDDS#3236]
The IDL documentation supports the use of "map" that corresponds to C++'s std::map object with the following language

`typedef map<T,V> M1;`

It appears that either TAO or OpenDDS (or both) don't support this type. Are there plans to implement this typing? And if not, what is the preferred solution when you want to create an IDL that has a property which is a list of key-value pairs? Do you suggest defining a struct for the key-value pairs and then using a sequence?

**Comment by mitza-oci** [comment]
> The IDL documentation supports the use of "map"

Does this refer to the OMG IDL spec or somewhere else?

> that corresponds to C++'s std::map object with the following language
> 
> `typedef map<T,V> M1;`
> 
> It appears that either TAO or OpenDDS (or both) don't support this type. Are there plans to implement this typing?

Right, it's not yet implemented.  I don't know of any active development but we can help if you'd like to sponsor work on it or contribute code.

> And if not, what is the preferred solution when you want to create an IDL that has a property which is a list of key-value pairs? Do you suggest defining a struct for the key-value pairs and then using a sequence?

Yes, the sequence-of-struct pattern has been traditionally used for this.  It's even in the DDS spec:s

```
    struct QosPolicyCount {
                QosPolicyId_t policy_id;
                long count;
                };

    typedef sequence<QosPolicyCount> QosPolicyCountSeq;
```
which is from the core DCPS spec, and 
```
    struct Property_t {
      string name;
      string value;
      boolean propagate;
    };
    typedef sequence<Property_t> PropertySeq;
```
which is used in RTPS and DDS Security.

**Comment by JonathanWheelerL3** [comment]
> Does this refer to the OMG IDL spec or somewhere else?

In OMG's IDL spec, yes. Specifically 7.4.13.4.3.1 in IDL v4.0 or higher (formal/2016-04-02). It is also found in section 6.31.1 of their C++11 Language v1.3 or higher (formal/2018-01-06).

> Right, it's not yet implemented. I don't know of any active development but we can help if you'd like to sponsor work on it or contribute code.

Can you explain what you mean by sponsor work on it? I'm afraid I'm a C# developer (hence why I've made my way here from OpenDDSharp) so I'm not keen on the idea of spinning up on C++ through this effort.

> Yes, the sequence-of-struct pattern has been traditionally used for this.

That's what I thought. Thanks for your help!

**Comment by mitza-oci** [comment]
> > Does this refer to the OMG IDL spec or somewhere else?
> 
> In OMG's IDL spec, yes. Specifically 7.4.13.4.3.1 in IDL v4.0 or higher (formal/2016-04-02). It is also found in section 6.31.1 of their C++11 Language v1.3 or higher (formal/2018-01-06).

Correct, there's a lot that's been specified that we don't implement yet.  This is documented in the Developer's Guide.


> > Right, it's not yet implemented. I don't know of any active development but we can help if you'd like to sponsor work on it or contribute code.
> 
> Can you explain what you mean by sponsor work on it? I'm afraid I'm a C# developer (hence why I've made my way here from OpenDDSharp) so I'm not keen on the idea of spinning up on C++ through this effort.

https://objectcomputing.com/products/opendds/opendds-consulting-and-support

**Comment by tmayoff** [comment]
If this isn't being worked on elsewhere I'd like to try to implement this myself, I'd just need some guidance on where to start with it.

**Comment by iguessthislldo** [comment]
@tmayoff You'd have to start by implementing it in [`tao_idl`](https://github.com/DOCGroup/ACE_TAO/tree/master/TAO/TAO_IDL). For that at least the following would have to happen:

- Implement the parsing of `map` types in IDL:
  - Get a copy of the latest OMG IDL spec. Right now that is [4.2](https://www.omg.org/spec/IDL/4.2/PDF). Map is defined in section 7.4.13.4.3.1 in that document.
  - Create a `AST_Map` class that inherits from `AST_Type` to represent a instance of a map type. [Here is the one for `seqeunce`](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/include/ast_sequence.h), which is probably the most similar.
  - Add a method for creating an instance of that class to [`AST_Generator`](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/include/ast_generator.h).
  - If you're not familiar with Bison/flex I'd suggest getting a least a little familiar. There are plenty of introductory tutorials online. 
    - Working on these files requires regenerating the commited files generated by Bison and flex. This can be done by running [`regen.pl`](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/regen.pl). This only works when makefiles can be used, so it's much easier to do on Linux or macOS than on Windows.
  - The `map` keyword has to be added to the [the flex file](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/fe/idl.ll).
    - Because map is IDL4-only, it would need a check to "reject" it in IDL3 [like this IDL4-only keyword](https://github.com/DOCGroup/ACE_TAO/blob/91d4aee881d200a00d5cdae63e47555609af66b9/TAO/TAO_IDL/fe/idl.ll#L160-L167), which allows it to be used as an identifier like for a type or member name. This prevents potentially breaking existing IDL unnecessarily.
  - The actual IDL grammar for map needs to be added to [the bison file](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/fe/idl.ypp). This will recognize when the user declared a map type in the IDL and turn it into a `AST_Map` to be added where ever the map type is being used. This is done by adding a grammar rule for the map, which again should be like [`sequence`](https://github.com/DOCGroup/ACE_TAO/blob/91d4aee881d200a00d5cdae63e47555609af66b9/TAO/TAO_IDL/fe/idl.ypp#L3892) and adding that rule to anywhere it can be used. You can reference the grammar given in the spec and compare it to uses of sequence in the bison file.
    - This isn't necessary to support map currently, but at some point it'd be ideal to add support for [map's "special case" annotations](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/docs/annotations.md#special-cases-of-annotations) inside the template expression like sequence already has.
- Implement the C++ code generation for map, how the IDL `map` a user writes gets turned into C++ code:
  -  How the mapping works might be a point of contention. The IDL spec is abstract and doesn't define what programming language types get generated from it. For this TAO follows the [IDL to C++ Language Mapping](https://www.omg.org/spec/CPP/1.3/PDF), but it doesn't mention `map` because this is spec is older than IDL4 and hasn't been updated since. Ideally it would use `std::map`, which is what the C++11 mapping @JonathanWheelerL3 mentioned specifies, but the original C++ mapping that TAO uses doesn't use standard C++ types for the other types. Other than that the only "`map` mapping" I know of is in the [DDS XTypes spec](https://www.omg.org/spec/DDS-XTypes/1.3/PDF) section 7.5.1.3.
  - For actually implementing it, I'm honestly less familiar with this side of things because most of my experience in tao_idl comes from implementing annotations which didn't need C++ code generation. However it would involve adding code to the [`be` (backend) parts of `tao_idl`](https://github.com/DOCGroup/ACE_TAO/tree/91d4aee881d200a00d5cdae63e47555609af66b9/TAO/TAO_IDL/be_include) and might also involve adding support code to `TAO/tao`.
- There are IDL 4 specific tests in [`TAO/tests/IDLv4`](https://github.com/DOCGroup/ACE_TAO/tree/master/TAO/tests/IDLv4). A new test should be added for maps. It should probably define various cases of maps in IDL and then try to use them in a C++ file.

There are probably some other things in I'm forgetting and random places that have to be updated, but that should be the gist of it. There could be more done to support maps in TAO itself like serialization, but CORBA doesn't have support for maps and even if it did, it'd probably be out of scope if we were adding support ourselves.

Once map is in TAO, support can be added in OpenDDS. This would involve updating `opendds_idl` for things like serialization and optionally adding support for this to the C++11 and Java mappings. I'm not going to elaborate on those at the moment because I probably just spent more time writing the TAO list than I should have.

**Comment by tmayoff** [comment]
Thanks for the detailed info, I'll see what I can get done

**Comment by mitza-oci** [comment]
Thanks for all the good info, @iguessthislldo and thanks for the offer to contribute, @tmayoff.  My suggestion would be to start with the first major bullet point "Implement the parsing of map types in IDL" and then we can discuss where to go next.

In case it's useful for anyone, I'll add on to...
> can be done by running [regen.pl](https://github.com/DOCGroup/ACE_TAO/blob/master/TAO/TAO_IDL/regen.pl). This only works when makefiles can be used, so it's much easier to do on Linux or macOS than on Windows.

It can run on Windows, you just need GNU Make for Windows along with flex and bison.  The makefile used for flex/bison doesn't try to compile any code, so `platform_macros.GNU` can be something like

```
use_pwd_call = 1
ECHO_BLANK_LINE = @echo -
GPERF = $(ACE_ROOT)\bin\ace_gperf
YACC = C:\your\path\to\bison
LEX = C:\your\path\to\flex
SHELL = cmd
```

**Comment by tmayoff** [comment]
Okay thanks, my plan was going to be to start with 1. adding a simple test for the IDL 2. bison/flex 3. then the AST_Generator stuff

**Comment by iguessthislldo** [comment]
I was hoping you'd chime in on how that worked on Windows Adam. I knew it was possible, but didn't know how.

> things need to be done in the order it was written in

Not particularity. The order of the list is the order I thought the dependencies went; test needs generated code, parser needs AST node, etc. You can do it in whatever order you want if it helps you.
